### ISSUE TITLE
docker issues

### ISSUE THREAD

**Issue by tomposmiko** [lxc/lxd#2255]
I'm not sure if it's OK if I list all docker issues here or better having them in separated issues.
I'm also not sure, if they are LXD problems or something in docker. But all of them come out in LXD (unprivileged) container.
- Distribution: Ubuntu 16.04 (host & guest)
- The output of "lxc info" or if that fails:
# lxc info

apicompat: 0
auth: trusted
environment:
  addresses:
- 10.128.0.102:8443
- 192.168.122.1:8443
  architectures:
- x86_64
- i686
  certificate: |
  [...]
  certificatefingerprint: 76d234347524b42545469a791f8414c7138a2fa2033ed999c3a8078141ae2b00
  driver: lxc
  driverversion: 2.0.3
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-21-generic
  server: lxd
  serverpid: 4291
  serverversion: 2.0.3
  storage: zfs
  storageversion: "5"
  config:
  core.https_address: '[::]'
  core.trust_password: true
  storage.zfs_pool_name: tank/lxd
  public: false
# Issue description
1. User cannot be changed via su in docker:

```
root@miko:~# docker run --rm -ti ubuntu
root@22b8ae247eff:/# adduser --shell /bin/bash --home /home/miko --gecos AA --disabled-password foo 
Adding user `foo' ...
Adding new group `foo' (1000) ...
Adding new user `foo' (1000) with group `foo' ...
Creating home directory `/home/miko' ...
Copying files from `/etc/skel' ...
root@22b8ae247eff:/# su - foo
Unable to cd to '/home/miko'
root@22b8ae247eff:/# 
```
1. apache cannot start

```
root@9bddfc56bf1a:/usr/local/apache2# apachectl -X
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 172.17.0.2. Set the 'ServerName' directive globally to suppress this message
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 172.17.0.2. Set the 'ServerName' directive globally to suppress this message
libgcc_s.so.1 must be installed for pthread_cancel to work
Aborted (core dumped)
```

I also experienced problem with mattermost, mysql and mongodb, however, cannot give you a detailed report at the moment.
As far as I remember it is an issue with volume mapping.

**Comment by kapilt** [comment]
i'm also running into issues running latest stable lxc (2.2) and latest stable docker 1.12.1 version info, i've tried different storage drivers on docker, haven't yet tried to intersect other storage drivers on lxd (its unclear what lxd init actually supports or if `dir` is actually an overlay). afaics all of @hallyn's patches for docker made it upstream.

```shell
  driver: lxc
  driverversion: 2.0.4
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-31-generic
  server: lxd
  serverpid: 17246
  serverversion: "2.2"
  storage: dir
  storageversion: ""
```

by default docker will run if told to explicitly not oom score adjust, but afaics none of the storage drivers but vfs will even start

btrfs
```
root@node-1:~# dockerd --storage-driver=btrfs --oom-score-adjust=0 
INFO[0000] libcontainerd: new containerd process, pid: 817 
WARN[0000] containerd: low RLIMIT_NOFILE changing to max  current=1024 max=65536
ERRO[0001] AppArmor enabled on system but the docker-default profile could not be loaded. 
FATA[0001] Error starting daemon: error initializing graphdriver: prerequisites for driver not satisfied (wrong filesystem?) 
```

aufs
```
root@node-1:~# dockerd --storage-driver=aufs --oom-score-adjust=0 
INFO[0000] libcontainerd: new containerd process, pid: 836 
WARN[0000] containerd: low RLIMIT_NOFILE changing to max  current=1024 max=65536
ERRO[0001] AppArmor enabled on system but the docker-default profile could not be loaded. 
FATA[0001] Error starting daemon: error initializing graphdriver: driver not supported 
```

overlayfs2
```
root@node-1:~# dockerd --storage-driver=overlay2 --oom-score-adjust=0 
INFO[0000] libcontainerd: new containerd process, pid: 857 
WARN[0000] containerd: low RLIMIT_NOFILE changing to max  current=1024 max=65536
ERRO[0001] AppArmor enabled on system but the docker-default profile could not be loaded. 
ERRO[0001] 'overlay' not found as a supported filesystem on this host. Please ensure kernel is new enough and has overlay support loaded. 
FATA[0001] Error starting daemon: error initializing graphdriver: driver not supported 
```

so running with vfs works (although this docker driver has loads of known issues)

```
root@node-1:~# dockerd --oom-score-adjust=0 
INFO[0000] libcontainerd: new containerd process, pid: 894 
WARN[0000] containerd: low RLIMIT_NOFILE changing to max  current=1024 max=65536
ERRO[0001] AppArmor enabled on system but the docker-default profile could not be loaded. 
ERRO[0001] 'overlay' not found as a supported filesystem on this host. Please ensure kernel is new enough and has overlay support loaded. 
INFO[0001] Graph migration to content-addressability took 0.00 seconds 
WARN[0001] Your kernel does not support swap memory limit. 
INFO[0001] Loading containers: start.                   
WARN[0001] Running modprobe bridge br_netfilter failed with message: modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/4.4.0-31-generic/modules.dep.bin'
modprobe: WARNING: Module bridge not found in directory /lib/modules/4.4.0-31-generic
modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/4.4.0-31-generic/modules.dep.bin'
modprobe: WARNING: Module br_netfilter not found in directory /lib/modules/4.4.0-31-generic
, error: exit status 1 
WARN[0001] Running modprobe nf_nat failed with message: `modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/4.4.0-31-generic/modules.dep.bin'
modprobe: WARNING: Module nf_nat not found in directory /lib/modules/4.4.0-31-generic`, error: exit status 1 
WARN[0001] Running modprobe xt_conntrack failed with message: `modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/4.4.0-31-generic/modules.dep.bin'
modprobe: WARNING: Module xt_conntrack not found in directory /lib/modules/4.4.0-31-generic`, error: exit status 1 
INFO[0001] Firewalld running: false                     
WARN[0001] Could not load necessary modules for IPSEC rules: Running modprobe xfrm_user failed with message: `modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/4.4.0-31-generic/modules.dep.bin'
modprobe: WARNING: Module xfrm_user not found in directory /lib/modules/4.4.0-31-generic`, error: exit status 1 
INFO[0001] Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address 

INFO[0001] Loading containers: done.                    
INFO[0001] Daemon has completed initialization          
INFO[0001] Docker daemon                                 commit=23cf638 graphdriver=vfs version=1.12.1
INFO[0001] API listen on /var/run/docker.sock           
```


but trying to run an image will then fail, with this on the client

```
root@node-1:~# docker run joshix/caddy
Unable to find image 'joshix/caddy:latest' locally
latest: Pulling from joshix/caddy

c367ae4f1784: Extracting [==================================================>] 4.886 MB/4.886 MB
docker: failed to register layer: ApplyLayer exit status 1 stdout:  stderr: permission denied.
See 'docker run --help'.
```


and this on the server

```
ERRO[0009] Handler for POST /v1.24/containers/create returned error: No such image: joshix/caddy:latest 
ERRO[0010] Error trying v2 registry: failed to register layer: ApplyLayer exit status 1 stdout:  stderr: permission denied 
ERRO[0010] Attempting next endpoint for pull after error: failed to register layer: ApplyLayer exit status 1 stdout:  stderr: permission denied 
```

basically failing to stack the runtime layer

**Comment by kapilt** [comment]
a little bit closer to working with docker 1.12.1 missing bit was setting both 
```
lxc config set node-1 security.nesting true
lxc config set node-1 security.privileged true
```

switched out zfs on lxd storage and aufs for docker

```dockerd --oom-score-adjust=0 --storage-driver=aufs```


running containers now errors with in ability to apply apparmor profile

```
# docker run joshix/caddy
Unable to find image 'joshix/caddy:latest' locally
latest: Pulling from joshix/caddy

c367ae4f1784: Pull complete 
Digest: sha256:17b83863ac097c481258ed163195b544ea17271b0b2ba5497a81401fda72d79f
Status: Downloaded newer image for joshix/caddy:latest
docker: Error response from daemon: oci runtime error: apparmor failed to apply profile: permission denied.
```

**Comment by stgraber** [comment]
@kapilt do you have the docker profile applied to your container? that profile includes a device entry that's supposed to take care of that very issue (by masking apparmor inside the container).

You may also want to try the docker 1.12 build from Ubuntu which includes all the needed patches for Docker to work inside LXD (with the docker profile applied):

**Comment by stgraber** [comment]
Download:
 - https://launchpad.net/ubuntu/+source/docker.io/1.12.1-0ubuntu10/+build/10759488/+files/docker.io_1.12.1-0ubuntu10_amd64.deb
 - https://launchpad.net/ubuntu/+source/runc/1.0.0~rc1-0ubuntu1/+build/10666130/+files/runc_1.0.0~rc1-0ubuntu1_amd64.deb
 - https://launchpad.net/ubuntu/+source/containerd/0.2.3-0ubuntu1/+build/10666152/+files/containerd_0.2.3-0ubuntu1_amd64.deb

Then:
```
root@ubuntu:~# dpkg -i *.deb
Selecting previously unselected package containerd.
(Reading database ... 25534 files and directories currently installed.)
Preparing to unpack containerd_0.2.3-0ubuntu1_amd64.deb ...
Unpacking containerd (0.2.3-0ubuntu1) ...
Preparing to unpack docker.io_1.12.1-0ubuntu10_amd64.deb ...
Unpacking docker.io (1.12.1-0ubuntu10) over (1.12.1-0ubuntu10) ...
Selecting previously unselected package runc.
Preparing to unpack runc_1.0.0~rc1-0ubuntu1_amd64.deb ...
Unpacking runc (1.0.0~rc1-0ubuntu1) ...
Setting up runc (1.0.0~rc1-0ubuntu1) ...
Setting up containerd (0.2.3-0ubuntu1) ...
Setting up docker.io (1.12.1-0ubuntu10) ...
Adding group `docker' (GID 116) ...
Done.
Processing triggers for ureadahead (0.100.0-19) ...
Processing triggers for systemd (229-4ubuntu8) ...
Processing triggers for man-db (2.7.5-1) ...
root@ubuntu:~# ls
containerd_0.2.3-0ubuntu1_amd64.deb  docker.io_1.12.1-0ubuntu10_amd64.deb  runc_1.0.0~rc1-0ubuntu1_amd64.deb
root@ubuntu:~# docker run joshix/caddy
docker: Cannot connect to the Docker daemon. Is the docker daemon running on this host?.
See 'docker run --help'.
root@ubuntu:~# cat /var/log/^C
root@ubuntu:~# systemctl -a | grep docker
  docker.service                               loaded    inactive   dead      Docker Application Container Engine
  docker.socket                                loaded    inactive   dead      Docker Socket for the API
root@ubuntu:~# systemctl status docker.service
● docker.service - Docker Application Container Engine
   Loaded: loaded (/lib/systemd/system/docker.service; disabled; vendor preset: enabled)
   Active: inactive (dead)
     Docs: https://docs.docker.com
root@ubuntu:~# systemctl start docker.socket
root@ubuntu:~# systemctl -a | grep docker
  docker.service                               loaded    inactive   dead      Docker Application Container Engine
  docker.socket                                loaded    active     listening Docker Socket for the API
root@ubuntu:~# docker run joshix/caddy
Unable to find image 'joshix/caddy:latest' locally
latest: Pulling from joshix/caddy
c367ae4f1784: Pull complete
Digest: sha256:17b83863ac097c481258ed163195b544ea17271b0b2ba5497a81401fda72d79f
Status: Downloaded newer image for joshix/caddy:latest
Activating privacy features... done.
http://:2015
```

**Comment by stgraber** [comment]
The above is inside a clean Ubuntu 16.04 unprivileged container, launched with:

    lxc launch ubuntu:x docker -p default -p docker

**Comment by stgraber** [comment]
@mwhudson any idea why I had to manually start docker.socket? ^

**Comment by mwhudson** [comment]
No. I've never had to do that. Any clues in the journal?

**Comment by kapilt** [comment]
@stgraber thanks! using the distro packages for 1.12.1 everything worked out of the box as unprivileged.

**Comment by stgraber** [comment]
@tomposmiko did you make any progress on this issue?

Your initial report indeed sounds like something going wrong with overlayfs/aufs on top of your zfs backed container. It seems unlikely to be something LXD can do anything about as it's not related to any of our security policies...

Might be interesting to make sure your kernel is up to date, your LXD too and maybe try it inside an Ubuntu 16.10 (ubuntu-daily:16.10) container to see if Docker 1.12 somehow changes that.
